### PR TITLE
Add initial tracing hooks to instrumentation.

### PR DESCRIPTION
**Pull request by ja30278**
For now, this only adds the stub implementation and some helper 'wrappers' so that we can disable functionality we don't plan to use initially and give us a place to hook custom functionality.

Also adds initial tests.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -19,6 +20,7 @@ module.exports = {
     this.errorReporter = ErrorReporter(pkg, env);
     this.metrics = Metrics(pkg, env);
     this.profiler = new Profiler(this, pkg, env);
+    this.tracer = Tracer(this, pkg, env);
```
Comment (dafortune) [inline index.js]: it is already there but using caps for not a constructor kind of breaks the convention :(

File: lib/stubs.js [hunk 2]
```diff
@@ -1,8 +1,24 @@
...
+  context: returnNull,
+  tracer: () => { return emptyTracer },
+  setOperationName: noop,
+};
+
+var emptyTracer = {
+  startSpan: () => { return emptySpan; },
+  inject: noop,
+  extract: returnNull
+}
```
Comment (dafortune) [inline lib/stubs.js]: missing semicolon; I think we should track adding a linter if we don't have it

File: lib/tracer.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    AUTH0_ENVIRONMENT: 'auth0.environment',
+    AUTH0_CHANNEL: 'auth0.channel'
+  }, tracing.Tags);
+
+  // Wrap the native opentracing 'span' object
+  // with a reduced API that discourages the use
+  // of features we don't want to use yet, while
+  // also giving us a place to hook additional
+  // functionality.
+  obj._wrapSpan = function(srcSpan) {
```
Comment (dafortune) [inline lib/tracer.js]: Why do we need to expose `_wrapSpan` into the public interface?

File: lib/tracer.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    const span = obj.startSpan(name, {
+      childOf: parentSpan
+    });
+    try {
+      fn(span);
+      span.finish();
+    } catch (e) {
+      span.setTag(obj.Tags.ERROR, true);
+      span.finish();
+      throw (e);
```
Comment (dafortune) [inline lib/tracer.js]: minor: it is not wrong, but it is not usual to put parens around thrown exceptions

File: test/stubs.test.js [hunk 5]
```diff
@@ -145,4 +146,46 @@ describe('stubs', function() {
 
   });
 
+  describe('tracer', function() {
```
Comment (dafortune) [inline test/stubs.test.js]: Why are we interested only in TypeError and not all errors regardless of the type? I mean, it does not looks like for an stub to throw in any case, unless it is a throwing stub :)

File: test/tracer.test.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
+const assert = require('assert');
+const opentracing = require('opentracing');
+const $require = require('proxyquire').noPreserveCache();
```
Comment (dafortune) [inline test/tracer.test.js]: It is fine but a pity having to use proxyquire for this, can't we just inject the dependency?

File: lib/tracer.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  const tracer = tracerFactory.create(agent, pkg, env);
+
+  const obj = {
+    _tracer: tracer,
+    _logger: agent.logger,
+    FORMAT_HTTP_HEADERS: tracing.FORMAT_HTTP_HEADERS,
+    FORMAT_TEXT_MAP: tracing.FORMAT_TEXT_MAP
+  };
+
+  obj.Tags = Object.assign({
```
Comment (dafortune) [inline lib/tracer.js]: Is this just an "enum" or it is supposed to change for every instance of the tracer? if it is just an enum we might want to put it outside so we calculate it only one time.

**Comment by dafortune** [comment]
CI failed

File: lib/tracer.js (hunk 7 above)
Comment (ja30278) [reply lib/tracer.js]: ✅

File: lib/tracer.js (hunk 4 above)
Comment (ja30278) [reply lib/tracer.js]: ✅

File: lib/tracer.js (hunk 3 above)
Comment (ja30278) [reply lib/tracer.js]: We don't. Fixed.

File: lib/stubs.js (hunk 2 above)
Comment (ja30278) [reply lib/stubs.js]: ✅

File: lib/tracer.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+      fn(span);
+      span.finish();
+    } catch (e) {
+      span.setTag(obj.Tags.ERROR, true);
+      span.finish();
+      throw (e);
+    }
+  };
+
+  obj.captureAsync = function(name, fn, parentSpan) {
```
[the remark this answers is not in the document]
Comment (ja30278) [reply lib/tracer.js]: Removed this until we have a better story for handling async things. There is a TracedPromise library that we could use/improve, but it seems to have some problems based on the github issues: https://github.com/lightstep/tracedpromise

File: test/tracer.test.js (hunk 6 above)
Comment (ja30278) [reply test/tracer.test.js]: ✅ See what you think about this.

**Comment by ja30278** [comment]
Also fixed the CI issue (which was specific to Node4).

File: test/stubs.test.js (hunk 5 above)
Comment (ja30278) [reply test/stubs.test.js]: I think this is mostly just to check that we don't remove an expected property from the stub, but Error also covers that, so 🤷‍♂️ I was honestly mostly cargo-culting what was already here. Changed my additions to 'Error'

**Pull request merged**